File: Sorting and Searching/Sum of Four Values/SumofFourValues.py
import sys
from time import time
s = time()
input = sys.stdin.readline

n,x = map(int, input().split())
a = list(map(int,input().split()))
for i in range(n):
    a[i] = (a[i],i+1)
a.sort()
for i in range(n-3):
    for j in range(i+1,n-2):
        target = x-a[i][0]-a[j][0]
        l,r = j+1, n-1
        while l<r:
            curr = a[l][0]+a[r][0]
            if curr==target :
                sys.stdout.write(str(a[i][1])+' '+str(a[j][1])+' '+str(a[l][1])+' '+str(a[r][1])+'\n')
                sys.stderr.write("Time taken: "+str(time()-s)+"s\n")
                sys.exit()
            elif curr<target:
                l += 1
            else:
                r -= 1

sys.stdout.write("IMPOSSIBLE\n")
sys.stderr.write("Time taken: "+str(time()-s)+"s\n")

File: Sorting and Searching/Sum of Four Values/SumofFourValues.cpp
#include<cstdio>
#include<algorithm>

using namespace std;

int main(){
    int n, x;
    scanf("%d %d", &n, &x);
    pair<int, int> a[n];
    for(int i=0;i<n;i++){
        scanf("%d", &a[i].first);
        a[i].second = i+1;
    }
    sort(a, a+n);
    for(int i=0;i<=n-3;i++){
        for(int j=i+1;j<=n-3;j++){
            int l = j+1, r = n-1;
            while(l<r){
                int curr = a[i].first + a[j].first + a[l].first + a[r].first;
                if(curr == x){
                    printf("%d %d %d %d\n", a[i].second, a[j].second, a[l].second, a[r].second);
                    return 0;
                }
                else if(curr < x) l++;
                else r--;
            }
        }
    }
    printf("IMPOSSIBLE\n");
}